File: datastructures_implemented/hash_tables/hash_table.py
from dataclasses import dataclass
from typing import Any


BUCKETS: int = 64


@dataclass
class Node:
    val: tuple[Any, Any]
    next: "Node | None" = None

class HashTable:
    def __init__(self):
        self.table = [None for _ in range(BUCKETS)]

    def _hash(self, s: str) -> int:
        return len(s) % BUCKETS

    def _get_index(self, s):
        if isinstance(s, str):
            return self._hash(s)
        return self._hash(str(s))

    def _extend_linked_list(self, head: Node, key, val):
        curr = head

        while curr.next is not None:
            curr = curr.next

        curr.next = Node((key, val))

    def add(self, key, value):
        idx = self._get_index(key)

        if self.table[idx] is None:
            self.table[idx] = Node((key, value))
        else:
            if isinstance(self.table[idx], Node):
                self._extend_linked_list(self.table[idx], key, value)

    def get(self, key):
        idx = self._get_index(key)
        curr = self.table[idx]

        while curr is not None:
            if curr.val[0] == key:
                return curr.val[1]
            curr = curr.next

        return None

    def iter(self):
        for bucket in self.table:
            curr = bucket

            while curr is not None:
                yield curr.val
                curr = curr.next


def main():
    hash_table = HashTable()

    hash_table.add("First", 1)
    hash_table.add("Second", 2)
    hash_table.add("Third", 3)

    print(hash_table.get("First"))
    print(hash_table.get("Third"))


if __name__ == "__main__":
    main()

File: datastructures_implemented/hash_tables/hash_table.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

const int BUCKETS = 64;


typedef struct 
{
        char* key;
        int value;
} Value;


typedef struct Node
{
        Value val;
        struct Node* next;
} Node;


typedef struct 
{
        Node hash_table[BUCKETS];
} HashTable;

HashTable hashtable_create(void);
void hashtable_add( char* key, int value );
int hashtable_get( char* key );
int hash( char* input );
int get_idx( char* input );
void extend_linked_list( Node* head ); 


int main() 
{
        return 0;
}

HashTable hashtable_create()
{
        HashTable hash_table;
        for (int i=0; i < BUCKETS; i++)
        {
                hash_table.hash_table[i].val = NULL;
                hash_table.hash_table[i].next = NULL;
        }
        return hash_table;
}

void hashtable_add(HashTable* hash_table, char* key, int value)
{
    int idx = get_idx(key);

    if (hash_table->hash_table[idx].val == NULL)
    {
        hash_table->hash_table[idx].val = malloc(sizeof(Value));

        hash_table->hash_table[idx].val->key = key;
        hash_table->hash_table[idx].val->value = value;
        hash_table->hash_table[idx].next = NULL;
    }
    else
    {

    }
}

int hashtable_get( char* key );

int hash( char* input )
{
        return strlen(input) % BUCKETS;
}

int get_idx( char* input )
{
        return hash( input );
}

void extend_linked_list( Node* head, char* key, int val )
{
        Node* curr = head;

        while ( curr->next != NULL )
        {
                curr = curr->next;
        }

        Value v = { key, val };
        Node n = { v, NULL };
        curr->next = &n;
}